### PR TITLE
Weights visualization

### LINKED ISSUE #18: Generate histograms of clipped/quantised model weight distributions
## Description

Generate histograms of original and quantised/clipped weight distributions for our baseline. Inspect codebase to see if this plotting code already exists, otherwise write a simple function.

## Checklist before resolving issue
- [ ] I have included the issue number in my PR description
- [ ] I have rebased/pulled in changes from `main` and fixed any conflicts
- [ ] If I have changed the models/training code, I have run `dry_run.sh` on a CUDA device successfully

### PR DESCRIPTION
**Pull request by heisenger**
## Description

Generate histograms of original and quantised/clipped weight distributions for our baseline. Inspect codebase to see if this plotting code already exists, otherwise write a simple function.

Closes #18 

## Checklist before resolving issue
- [x] I have included the issue number in my PR description
- [x] I have rebased/pulled in changes from `main` and fixed any conflicts
- [x] If I have changed the models/training code, I have run `dry_run.sh` on a CUDA device successfully

### CODE REVIEW

File: visualizations/weight_distribution.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import torch
+import matplotlib.pyplot as plt
+import numpy as np
+import argparse
+import os
+
+
+def plot_histogram(data, title, save_path):
+    plt.figure(figsize=(8, 5))
```
Comment (BrownianNotion) [inline visualizations/weight_distribution.py:9]: What does the plot look like without specifying figsize? Just thinking it might be better to let matplotlib choose the dimensions

File: visualizations/weight_distribution.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    plt.hist(data, bins=100, alpha=0.7)
```
Comment (BrownianNotion) [inline visualizations/weight_distribution.py:10]: Adjust the bins/x-axis scaling - right now almost all of it is blank space and not informative. See Figure 3 from BitDistiller, can use similar ticksizes to them.

File: visualizations/weight_distribution.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    # Load the .pt file
+    file_path = args.file_path
+    data = torch.load(file_path, map_location="cpu")  # Load onto CPU
+
+    # Ensure save directory exists
+    os.makedirs(args.save_dir, exist_ok=True)
+
+    # Plot all weights histogram
+    weights = []
+    if "clip" in data and isinstance(data["clip"], list):
```
Comment (BrownianNotion) [inline visualizations/weight_distribution.py:43]: Double checking are all weights from `.pt` from the `qlinear` layers? We should only be plotting these as these are the quantised ones (eg. as opposed to attention weights).

File: visualizations/weight_distribution.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                        weights.append(tensor.view(-1).numpy())  # Flatten
+
+    flattened_weights = np.concatenate(weights).tolist()
+    plot_histogram(
+        flattened_weights,
+        "Histogram of All Extracted Tensor Values",
+        os.path.join(args.save_dir, f"{args.plot_name}_all_weights.png"),
+    )
+
+    # Plot final down projection layer weight histogram
```
Comment (BrownianNotion) [inline visualizations/weight_distribution.py:58]: Why are the down proj weights treated separately?

**Comment by BrownianNotion** [comment]
Also, not sure why the description didn't use the PR template which should have this + a checklist, but can you add the issue number with #[no.]

File: visualizations/weight_distribution.py (hunk 2 above)
Comment (BrownianNotion) [inline visualizations/weight_distribution.py:10]: Nit: specify `edgecolor='white'` in `plt.hist` for better visual separation of bins

**Comment by fredericowieser** [comment]
This looks pretty good :) I think @BrownianNotion covers

The figures you have included as .pngs are those 1bit or 2bit or... weights?

**Review by fredericowieser — APPROVED** [review]

File: visualizations/weight_distribution.py (hunk 2 above)
Comment (heisenger) [reply visualizations/weight_distribution.py:10]: sounds good added this

Comment (heisenger) [reply visualizations/weight_distribution.py:10]: i now change the y-axis to be log scale, it looks a lot clearer now hopefully

File: visualizations/weight_distribution.py (hunk 1 above)
Comment (heisenger) [reply visualizations/weight_distribution.py:9]: think it looks clearer - see updated version

File: visualizations/weight_distribution.py (hunk 4 above)
Comment (BrownianNotion) [reply visualizations/weight_distribution.py:58]: ahhh that's interesting, any ideas why? non-blocking so we can merge in, but would be good to know.

**Review by BrownianNotion — APPROVED** [review]
Looks good! would be good to know why the authors only plot the final layer down proj but this is non-blocking.

**Comment by BrownianNotion** [comment]
Maybe just rename the example files to include the number of bits + quantisation method. Unclear right now as @fredericowieser mentioned
